### PR TITLE
feat(app-platform): prevent users from creating sentry apps with greater scopes than they have

### PR DESCRIPTION
**Pull request by scefali**
This PR prevents users from creating sentry apps, or modifying existing sentry apps, to have scopes greater than the user has. This closes the existing bug where a user can escalate their permissions through sentry apps. We already have a change that prevents users from viewing tokens/secrets of existing apps that have scopes greater than the user possesses.

I have chosen to do the check on scopes in the `SentryAppSerializer` since it's the easiest place to make the change. However, one potential issue is that we return a `400` instead of a `403` when the user tries to increase scopes beyond what their user possesses. But maybe that's OK.

One decision I had to make was whether to let the user be able to choose scopes above their user in the UI. I decided to let them do that because I want users to know that it is possible to make an app with those scopes, but they would have to contact an administrator to make the change for them as explained in the error. Also, I had to make a small UI adjustment so the dropdown appears on top of the error message which allows the user to select a different choice to clear the error. See below:

Screen Shot 2019-10-08 at 4 44 00 PM

### CODE REVIEW

File: src/sentry/api/serializers/rest_framework/sentry_app.py [hunk 1]
```diff
@@ -113,6 +118,27 @@ def validate_allowedOrigins(self, value):
...
+
+        validation_errors = []
+        for scope in value:
+            # if the existing instance already has this scope, skip the check
+            if self.instance and self.instance.has_scope(scope):
+                continue
+            # add an error if the requester lacks permissions being requested
+            if not self.access.has_scope(scope):
+                validation_errors.append(
+                    "Requested permission of %s exceeds requester's permission. Please contact an administrator to make the requested change."
```
Comment (MeredithAnya) [inline src/sentry/api/serializers/rest_framework/sentry_app.py]: `Please contact an administrator to make the requested change` is this supposed to me requested change of the role permissions? or do you mean like the actually changing of the scope?

Might be enough to say `Insufficient permissions: Your role does not have access to %s permission`

**Review by MeredithAnya — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/less/integrations.less [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.permission-selection .form-field-error-reason {
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/static/sentry/less/integrations.less]: @MeredithAnya Sure, I will talk to some of the front end folks about a better way of doing this.

File: src/sentry/api/serializers/rest_framework/sentry_app.py (hunk 1 above)
Comment (scefali) [reply src/sentry/api/serializers/rest_framework/sentry_app.py]: @MeredithAnya I want the bit about contacting an administrator to let the user know that using that scope is possible but they need an admin to do it for them. My concern about saying `Insufficient permissions: Your role does not have access to %s permission` is that they might not realize there is a way to get past this if they really need to. What do you think?

File: src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/sentryApplicationDetails.tsx [hunk 3]
```diff
@@ -31,8 +31,33 @@ import {
...
+    const allScopes = allChoices.reduce(
+      (_allScopes: string[], choice) => _allScopes.concat(_.get(choice, 'scopes', [])),
+      []
+    );
+
+    if (allScopes.includes(scope)) {
+      return permObj.resource as Resource;
+    }
+  }
+  return undefined;
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/sentryApplicationDetails.tsx]: @MeredithAnya I think `undefined` is more appropriate. This function is similar to a ` find` function which returns `undefined` if you can't find the element you're searching for. That being said, the way the code is written `null` would produce the same end behavior.

File: src/sentry/api/serializers/rest_framework/sentry_app.py (hunk 1 above)
Comment (MeredithAnya) [reply src/sentry/api/serializers/rest_framework/sentry_app.py]: yeah I guess we can keep it. seems a little wordy to me but can't really think of anything better so...

**Review by MeredithAnya — APPROVED** [review]

**Pull request merged**